### DISCUSSION TITLE
Implement Profile Table and Endpoints for Users in ASP.Net Web API Project

### DISCUSSION THREAD

**Discussion by johnsmith0305** [KGG-Dev-Course/taskmaster-api#18] [Ideas]
### Objective:
To enhance the user management system by implementing a profile table and relevant endpoints in the ASP.Net Web API project. This enhancement will allow users to store additional information beyond the basic user details provided by the IdentityUser.

### Details:
Profile Table:
- Create a new table named **"UserProfile"** or similar to store additional user information.
- Define the necessary fields such as **UserId** (foreign key referencing IdentityUser), **FirstName**, **LastName**, **Email**, **PhoneNumber**, etc.
- Implement proper relationships between the **IdentityUser** and **UserProfile** tables.

Endpoints:
1. **Create Profile:**
- Endpoint: **'/api/Profile'**
- Method: POST
- Description: Allows users to create their profile by providing necessary details like first name, last name, email, phone number, etc.

2. **Get Profile by User ID:**
- Endpoint: **'/api/Profile/{userId}'**
- Method: GET
- Description: Fetches the profile information for a specific user based on their ID.

3. **Update Profile**
- Endpoint: **'/api/Profile/{uesrId}'**
- Method: PUT
- Description: Allows users to update their profile information with new details.

4. **Delete Profile**
- Endpoint: **/api/Profile/{id}'**
- Method: DELETE
- Description: Allows user to delete their profile permanently.

### Additional Considerations:
- Ensure proper validation for data input.
- Implement authentication and authorization mechanisms for profile-related endpoints.
- Document API endpoints using Swagger or similar tools for ease of understanding and usage. (automatic)
- Write unit tests to ensure the functionality and security of profile-related operations. (optional)

### Implementation Plan:
1. Define the structure of the UserProfile table and its relationship with IdentityUser.
2. Implement CRUD operations for profile management.
3. Integrate authentication and authorization mechanisms for profile endpoints.
4. Write unit tests to validate the functionality. (optional)
5. Document API endpoints using Swagger. (automatic)

**Comment by gitfullstacker** [comment]
@johnsmith0305 Hi, why is it moved into discussion?

**Comment by johnsmith0305** [comment]
@chenxidev1129 @kgg0305 I'd love to hear your thoughts on creating a new table for user profile.

**Reply by kgg0305** [reply]
Sounds good

**Reply by fullstackdev0305** [reply]
@johnsmith0305 That's a great template for a ticket.

**Reply by gitfullstacker** [reply]
@fullstackdev0305 Thank for paticipation.

**Reply by gitfullstacker** [reply]
@fullstackdev0305 I hope we're implementing your guidance in this project.

**Comment by gitfullstacker** [comment]
I believe we should include the following fields for the new tables:

- **UserId** String (FK)
- **Photo** String
- **FirstName** String
- **LastName** String
- **PhoneNumber** String
- **AboutMe** String

**Reply by kgg0305** [reply]
I think we need these fields more:
- **Gender** Bit
- **DateOfBirth** String

**Comment by fullstackdev0305** [comment]
We can consider implementing a step form for the user profile.
https://primeng.org/steps/seat

**Reply by johnsmith0305** [reply]
Absolutely, we can utilize it when the form data becomes extensive or complex.

**Reply by johnsmith0305** [reply]
let's begin by defining the data structure.

**Comment by fullstackdev0305** [comment]
We should consider scenarios where a user might have multiple roles, such as being an admin and a regular user simultaneously.
In such a scenario, could there be multiple profiles associated with this user?

**Reply by kgg0305** [reply]
If that's the case, we'll allocate one profile per user.

**Comment by gitfullstacker** [comment]
I updated my quote like following:

- **UserId** String (FK)
- **FirstName** String
- **LastName** String
- **AboutMe** String
- **Gender** Bit
- **DateOfBirth** String
- **PhoneNumber** String
- **Photo** String

This is just for personal profile, so it is same whatever user is admin or user.

**Reply by johnsmith0305** [reply]
I agree

**Reply by fullstackdev0305** [reply]
Alright
